@here/harp-datasource-protocol/lib/Expr.ts

```diff
@@ -53,6 +53,11 @@ export class ExprDependencies {
      * `true` if the expression depends on the feature state.
      */
     featureState?: boolean;
+
+    /**
+     * `true` if this expression cannot be cached.
+     */
+    volatile?: boolean;
```

Comment:
Hmm. I have problem with this particular definition of `volatile`.

It actually should read as <<results are dependent on ''dynamic-properties'>> .... but unless i'm missing something, they can be cached very effectively.

In particular `volatile` currenty only expressions with `dynamic-properties` are volatile now, but apparently they can be effectively cached as long as no one calls `MapView.setDynamicProperty` ... and that's easily handled by `MemoCall` already.

I have bigger problem with `$id` and `feature-state` (not marked as volatile?) this is changes so often that i would say it's not feasible to cache them ...

So, summarizing i think that defining `volatile` here is bogus. Only TileGeometryCreator and MapView khow envs will change (per frame / per feature) so only they can know if something is really volatile or not.

Reply:
well, `volatile` is general and it may be used for all the current and future `Expr` that cannot be cached.. e.g. say we add an operator `["debug", expr]` that evaluates expr, prints its value and then returns it. well, in that case `["debug", expr]` should be marked as `volatile`.

`volatile` can also be used in cases where the engine is not yet optimizing the fetching of feature properties.  The current implementation of `setDynamicProperty`/`["dynamic-properties"]` is modifying the `Env` directly and there is no way to detect changes there yet. So at least for now we don't optimize it and mark it as `volatile`, this may change very soon btw.

about `["feature-state"]`, feature-state is creating a completely new property `$state` for each feature and `MemoExpr` will catch that. same for `["id"]` it rely on $id and that changes for each feature.